### PR TITLE
ipv4 support

### PR DESCRIPTION
**Pull request by furkansahin**
IPv4 Support
This PR enhances the network functionality by introducing IPv4 support.
In order to utilize IPv4, it is required to obtain an IPv4 subnet. Once
the IPv4 address is acquired, the host pulls this information from
Hetzner, and IP addresses are randomly assigned to the devices. In the
future, we can further improve this process by allowing customers to
specify the desired IP address.  If it is not feasible to assign an IP
address due to all addresses being in use, the system will gracefully
proceed without assigning one.

To enable local redirection, the PR leverages the local address space
169.254.*.*. This ensures efficient communication within the local
network while utilizing private IP addresses.

To try the patch,
1. You need to have a VmHost and order an ipv4 block from Hetzner for
that server. If you already have a host, you can just add an ipv4 block
and continue to step 3
2. Once you have both, add the vm_host to clover.
3. Once it's waiting, you can add the ipv4 address space you want to use
by using vm_host.create_addresses.
4. Create a VM for that host.
5. Fetch the ipv4 address and ssh in!

```
st = Prog::Vm::HostNexus.assemble("host_ip")
st.run 60

# make sure it's in wait state
st.vm_host.reload.create_addresses([{:ip_address=>"such as 1.1.1.0/24",
:source_host_ip=>"host_ip", :is_failover=>true/false}])

# check Addresses
st.vm_host.reload.assigned_subnets

# create a VM
mykey = File.read("path/to/your/key")
st = Prog::Vm::Nexus.assemble(mykey, nil)
st.run 70

# once it's in wait state, fetch the ipv4 address
st.vm.reload.ephemeral_net4.to_s
```
Now you can use it to ssh in.
Notes:
1. You have to have a subnet of at least /31. It doesn't work with
single ip ranges because we are assigning 1 ip to tap device and 1 ip
to the VM.
2. Reboot is an issue for the whole networking stack, same here. We'll
have to fix it.
3. We collect information of `is_failover_ip` but never use it now.
That will be useful when we get into ip block reassignments (e.g.
failover).

### CODE REVIEW

**Comment by fdr** [comment]
Seems like maybe you can obsolete this bucket idea with v4 link local addresses maybe?

File: model/vm_host.rb [hunk 1]
```diff
@@ -67,6 +67,30 @@ def ip6_random_vm_network(prefix = 79)
     end
   end
 
+  def ip4_random_vm_network
+    subnet = NetAddr::CIDR.create(ipv4_bucket)
+    ip_list = subnet.enumerate
+  
+    # remove the first and the last IP from the range
```
Comment (serprex) [inline model/vm_host.rb:74]: Purpose of this comment is obvious from code. Not obvious: why you want to remove the ends of the range

File: model/vm_host.rb [hunk 2]
```diff
@@ -67,6 +67,30 @@ def ip6_random_vm_network(prefix = 79)
...
+    subnets = ip_list.each_slice(4).map do |slice|
+      # Create a /30 subnet for each slice of 4 IPs.
+      # The CIDR.create method will raise an error if the slice does not contain exactly 4 IPs, so we need to check for this.
+      if slice.size == 4
+        NetAddr::CIDR.create("#{slice.first}/30")
+      end
+    end.compact
+  
+    # pick a random subnet
+    selected_subnet = subnets[SecureRandom.random_number(subnets.length)]
```
Comment (serprex) [inline model/vm_host.rb:88]:
```suggestion
    selected_subnet = subnets.sample
```

File: model/vm_host.rb [hunk 3]
```diff
@@ -67,6 +67,30 @@ def ip6_random_vm_network(prefix = 79)
...
+  def ip4_random_vm_network
+    subnet = NetAddr::CIDR.create(ipv4_bucket)
+    ip_list = subnet.enumerate
+  
+    # remove the first and the last IP from the range
+    ip_list.shift
+    ip_list.pop
+  
+    # split the remaining IPs into subnets of 4 IPs each
+    subnets = ip_list.each_slice(4).map do |slice|
```
Comment (serprex) [inline model/vm_host.rb:79]:
```suggestion
    subnets = ip_list.each_slice(4).filter_map do |slice|
```
removes need to call `.compact` at end

File: model/vm.rb [hunk 4]
```diff
@@ -61,6 +62,10 @@ def mem_gib
     product.cores * mem_gib_ratio
   end
 
+  def ip4
+    ephemeral_net4 == "" ?  "" : NetAddr::IPv4Net.parse(ephemeral_net4).network.to_s
```
Comment (furkansahin) [inline model/vm.rb:66]: I hate this, I'll try to fix it.

File: model/vm_host.rb [hunk 5]
```diff
@@ -67,6 +67,25 @@ def ip6_random_vm_network(prefix = 79)
...
+    return nil unless ipv4_bucket
+
+    subnet = NetAddr::IPv4Net.parse(ipv4_bucket)
+    selected_subnet = NetAddr::IPv4Net.new(subnet.nth(SecureRandom.random_number(subnet.len - 1).to_i), NetAddr::Mask32.new(32))
+    
+    ip4_random_vm_network if selected_subnet.network.to_s.nil? || vms.any? { |vm| vm.ip4 == selected_subnet.network.to_s }
+    selected_subnet
+  end
+
+  def vetho_ip4_local_random_vm_network
```
Comment (furkansahin) [inline model/vm_host.rb:80]: much better name needed here. Essentially, we get 1 address here and use 2 consecutive next addresses to assign route for vetho*, vethi* and tap*. So, we are splitting the local address space to three but only store 1 of them for the existence check.

File: prog/vm/nexus.rb [hunk 6]
```diff
@@ -82,10 +82,18 @@ def private_subnets
...
+  def q_net6
     vm.ephemeral_net6.to_s.shellescape
   end
 
+  def q_net4
+    vm.ip4.to_s.shellescape
+  end
+
+  def local_ipv4
+    vm.local_vetho_ip ? vm.local_vetho_ip.to_s.shellescape : ""
```
Comment (furkansahin) [inline prog/vm/nexus.rb:94]: hate this, gotta fix it.

File: prog/vm/nexus.rb [hunk 7]
```diff
@@ -176,6 +188,20 @@ def create_ipsec_tunnel(my_subnet, dst_vm, dst_subnet)
 
   def run
     host.sshable.cmd("sudo systemctl start #{q_vm}")
+    hop :setup_sshable
+  end
+
+  def setup_sshable
+    Sshable.create(host: vm.ephemeral_net6.network.to_s + "2") { _1.id = vm.id }
```
Comment (furkansahin) [inline prog/vm/nexus.rb:195]: not sure if this is a must, I may endup implementing a utility function to ssh into the vm instead of creating an sshable.

File: rhizome/lib/vm_setup.rb [hunk 8]
```diff
@@ -37,22 +37,16 @@ def vp
     @vp ||= VmPath.new(@vm_name)
   end
 
-  def prep(unix_user, public_key, private_subnets, gua, boot_image, max_vcpus, cpu_topology, mem_gib)
+  def prep(unix_user, public_key, private_subnets, gua, ip4, local_ip4, boot_image, max_vcpus, cpu_topology, mem_gib)
     interfaces
-    routes(gua, private_subnets)
+    routes6(gua, private_subnets)
+    routes4(ip4, local_ip4) unless ip4 == "''"
```
Comment (furkansahin) [inline rhizome/lib/vm_setup.rb:43]: this is disgusting, too, the parameter translation is an issue here. Moving them via json, parsing, etc. Gotta find something else, maybe we would allow nil here.

File: rhizome/lib/vm_setup.rb [hunk 9]
```diff
@@ -37,22 +37,16 @@ def vp
...
     install_systemd_unit(max_vcpus, cpu_topology, mem_gib)
     forwarding
   end
 
-  def network(unix_user, public_key, private_subnets, gua)
-    interfaces
-    routes(gua, private_subnets)
-    cloudinit(unix_user, public_key, private_subnets)
-    forwarding
-  end
```
Comment (furkansahin) [inline rhizome/lib/vm_setup.rb:54]: never used

File: rhizome/lib/vm_setup.rb [hunk 10]
```diff
@@ -73,17 +67,17 @@ def purge
...
     # a vexing bug whereby a freshly created link will, at least once,
     # spontaneously change its MAC address sometime soon after
     # creation, as caught by instrumenting reads of
     # /sys/class/net/vethi#{q_vm}/address at two points in time.  The
     # result is a race condition that *sometimes* worked.
-    r "ip link add vetho#{q_vm} addr #{gen_mac.shellescape} type veth peer name vethi#{q_vm} addr #{gen_mac.shellescape} netns #{q_vm}"
+    r "ip link show vetho#{q_vm} > /dev/null 2>&1 || ip link add vetho#{q_vm} addr #{gen_mac.shellescape} type veth peer name vethi#{q_vm} addr #{gen_mac.shellescape} netns #{q_vm}"
 
-    r "ip -n #{q_vm} tuntap add dev tap#{q_vm} mode tap user #{q_vm}"
+    r "ip netns exec #{q_vm} ip link show tap#{q_vm} > /dev/null 2>&1 || ip -n #{q_vm} tuntap add dev tap#{q_vm} mode tap user #{q_vm}"
```
Comment (furkansahin) [inline rhizome/lib/vm_setup.rb:80]: The changes here are my effort to make the function idempotent.

File: prog/vm/nexus.rb [hunk 11]
```diff
@@ -112,7 +120,9 @@ def allocate
 
   def start
     vm_host_id = allocate
-    vm.update(vm_host_id: vm_host_id, ephemeral_net6: VmHost[vm_host_id].ip6_random_vm_network.to_s)
+    vm.update(vm_host_id: vm_host_id, ephemeral_net6: VmHost[vm_host_id].ip6_random_vm_network.to_s,
+      ephemeral_net4: VmHost[vm_host_id].ip4_random_vm_network.to_s,
+      local_vetho_ip: VmHost[vm_host_id].vetho_ip4_local_random_vm_network.to_s)
```
Comment (enescakir) [inline prog/vm/nexus.rb:125]: I think it sends 3 different SQL query.
```suggestion
    vmh = VmHost[vm_host_id]
    vm.update(vm_host_id: vm_host_id, ephemeral_net6: vmh.ip6_random_vm_network.to_s,
      ephemeral_net4: vmh.ip4_random_vm_network.to_s,
      local_vetho_ip: vmh.vetho_ip4_local_random_vm_network.to_s)
```

File: rhizome/lib/vm_setup.rb (hunk 10 above)
Comment (fdr) [reply rhizome/lib/vm_setup.rb:81]: Interesting. I think "r" does return a string, I wonder if you could do better. Customarily, I check error messages from `ip netns add`  (or whatever), see `rescue CommandFail` elsewhere. I think this might give you a better result than embedded shell programs, and more efficient too (fewer forks, pipes, etc)

File: config.rb [hunk 12]
```diff
@@ -57,4 +57,7 @@ def self.development?
   def self.test?
     Config.rack_env == "test"
   end
+
+  override :hetzner_user, string
+  override :hetzner_password, string
```
Comment (fdr) [inline config.rb:62]: should have a clear: true on this. Also are you sure you mean override or 'optional'?

**Comment by fdr** [comment]
192 is a space customers are likely to use. How about a counter-proposal: the reserved region in `240.0.0.0/4`? AWS is squatting on it for some of their hops, I doubt they're the only ones, and it was penciled for a Class E networks a generation ago, and as slow as IPv6 adoption has been...I don't think the reservation status is ever gunna change.

Alternatively, can you use link local addresses for this, 169.254.0.0/16? I don't know if it conflicts with other concerns like DHCP in a meaningful way.

File: prog/vm/nexus.rb [hunk 13]
```diff
@@ -176,8 +191,8 @@ def create_ipsec_tunnel(my_subnet, dst_vm, dst_subnet)
 
   def run
     host.sshable.cmd("sudo systemctl start #{q_vm}")
-    vm.update(display_state: "running")
-    hop :wait
+    hop :wait if host.sshable.cmd("sudo systemctl is-active #{q_vm}")
```
Comment (furkansahin) [inline prog/vm/nexus.rb:194]: instead of this, I want to perform an ssh access check actually.

Comment (fdr) [reply prog/vm/nexus.rb:194]: what do you mean?

Comment (furkansahin) [reply prog/vm/nexus.rb:194]: When we switch the vm to running state, the client still cannot ssh-in a couple of more seconds. I realized in Ozgun's demo that the couple of ssh requests failed and once the vm is up and running properly, we were able to. So, I would like to change it here so that we only switch to running after verifying that the vm is sshable.

File: migrate/011_add_ipv4.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    alter_table(:vm_host) do
+      add_column :ipv4_bucket, :text, collate: '"C"'
+    end
+    alter_table(:vm) do
+      add_column :ephemeral_net4, :text, collate: '"C"'
```
[the remark this answers is not in the document]
Comment (furkansahin) [reply migrate/011_add_ipv4.rb]: removed it altogether

File: rhizome/lib/vm_setup.rb (hunk 10 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb]: I changed this but didn't test it yet, I'll do so, soon.

File: migrate/011_add_ipv4.rb [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+    end
+    create_table :assigned_address do
+      column :id, :uuid, primary_key: true, default: Sequel.lit("gen_random_uuid()")
+      column :cidr, :cidr, null: false, unique: true
+      foreign_key :address_id, :address, type: :uuid, null: false
+      foreign_key :dst_vm_id, :vm, type: :uuid, null: false
+    end
+    create_table :host_address do
+      column :id, :uuid, primary_key: true, default: Sequel.lit("gen_random_uuid()")
+      column :cidr, :cidr, null: false, unique: true
```
Comment (enescakir) [inline migrate/011_add_ipv4.rb:29]: Is it same with `.address.cidr`

File: migrate/011_add_ipv4.rb [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    end
+    alter_table(:vm_host) do
+      add_column :provider, :text, collate: '"C"'
+    end
+    alter_table(:vm) do
+      add_column :local_vetho_ip, :text, collate: '"C"'
+    end
+    create_table :address do
+      column :id, :uuid, primary_key: true, default: Sequel.lit("gen_random_uuid()")
+      column :provider, :text, collate: '"C"', null: false
```
Comment (enescakir) [inline migrate/011_add_ipv4.rb:16]: Is it same with `.host.provider`

File: prog/vm/nexus.rb [hunk 17]
```diff
@@ -176,8 +190,12 @@ def create_ipsec_tunnel(my_subnet, dst_vm, dst_subnet)
 
   def run
     host.sshable.cmd("sudo systemctl start #{q_vm}")
```
Comment (enescakir) [inline prog/vm/nexus.rb:192]: Is `systemctl start` idempotent? Or fail if service is already running?

File: prog/vm/nexus.rb [hunk 18]
```diff
@@ -176,8 +190,12 @@ def create_ipsec_tunnel(my_subnet, dst_vm, dst_subnet)
...
-    vm.update(display_state: "running")
-    hop :wait
+    if host.sshable.cmd("sudo systemctl is-active #{q_vm}")
+      vm.update(display_state: "running")
+      hop :wait
+    end
+
+    nap 5
```
Comment (enescakir) [inline prog/vm/nexus.rb:198]: I think this part can be a separate PR

File: prog/vm/host_nexus.rb [hunk 19]
```diff
@@ -1,20 +1,33 @@
...
-  def self.assemble(sshable_hostname, location: "hetzner-hel1")
+  subject_is :vm_host
+
+  def self.assemble(sshable_hostname, provider: nil, location: "hetzner-hel1", hetzner_server_number: nil)
     DB.transaction do
       sa = Sshable.create(host: sshable_hostname)
-      VmHost.create(location: location) { _1.id = sa.id }
-
+      VmHost.create(location: location, provider: provider) { _1.id = sa.id }
+      HetznerHost.create(server_number: hetzner_server_number) { _1.id = sa.id } if provider == "hetzner"
```
Comment (enescakir) [inline prog/vm/host_nexus.rb:10]: I saw `"hetzner"` string several time. It might be a good candidate for a constant.

File: migrate/011_add_ipv4.rb [hunk 20]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    alter_table(:sshable) do
+      set_column_allow_null :host
+    end
+    alter_table(:vm_host) do
+      add_column :provider, :text, collate: '"C"'
```
[the remark this answers is not in the document]
Comment (furkansahin) [reply migrate/011_add_ipv4.rb:9]: I think splitting the location to provider and region/location makes more sense. Especially because provider can be used to filter other stuff in our code such as which endpoint to hit as it is here.

File: migrate/011_add_ipv4.rb (hunk 15 above)
Comment (furkansahin) [reply migrate/011_add_ipv4.rb:29]: Looking into how Mars gave us the servers with 1 subnet that has the server ip in it, I now think that, no they are not. Address will have /~32 subnet while host_address will be /32

File: config.rb [hunk 21]
```diff
@@ -64,4 +64,7 @@ def self.test?
   override :timeout, 10, int
   override :versioning, false, bool
   override :allowed_vm_host_users, "", array(string)
+  optional :hetzner_user, string, clear: true
```
[the remark this answers is not in the document]
Comment (furkansahin) [reply config.rb:67]: That's an assumption, yes. @fdr what do you think about this?

File: prog/vm/nexus.rb (hunk 13 above)
Comment (furkansahin) [reply prog/vm/nexus.rb]: removed for now since it's not related to this pr

File: prog/vm/host_nexus.rb (hunk 19 above)
Comment (fdr) [reply prog/vm/host_nexus.rb]: I'd be careful how I make the constant,  it'd be nice if printing it out was homoiconic, e.g.:

```
[2] clover-development(main)> module Provider
[2] clover-development(main)*   module Hetzner
[2] clover-development(main)*   end  
[2] clover-development(main)* end  
=> nil
[3] clover-development(main)> Provider::Hetzner
=> Provider::Hetzner
```

Example of something less useful:

```
[1] clover-development(main)> module Provider
[1] clover-development(main)*   Hetzner = 'hetzner'
[1] clover-development(main)* end  
=> "hetzner"
[2] clover-development(main)> Provider::Hetzner
=> "hetzner"
```

The reason being, you will always see the result of evaluation should you read any representation of data including the value.

Finally, I have found practical problems trying to avoid some kind of data literal for these kinds of things, in that the amount of representation swapping with text out of database can make it a cure worse than the disease. It also tends to prove strange if you serialize or de-serialize JSON.  Consider this problem, using the first variant of a `Provider::Hetzner` module:

```
puts JSON.generate({provider: Provider::Hetzner})
{"provider":"Provider::Hetzner"}
```

And here's what happens in a round-trip between generating and parsing:

```
puts JSON.parse(JSON.generate({provider: Provider::Hetzner}))
{"provider"=>"Provider::Hetzner"}
```

Now...what are you going to do with that string? Write `case` statements in one or even a few places to convert it into a `Provider::Hetzner` value?  Use eval? No, don't do that, it's a security vulnerability! How about `const_get`, as we are more or less obliged to use for Progs to match database record contents to code? I guess you can do that. But you still have to watch out for what you are const-getting. It's for this reason that `Prog::`s are boxed in their own module, to make it a little harder to wiggle an arbitrary code execution into the database, rather than const_getting absolutely anything. We could probably tighten it up even more.

But by the time you're done, was it worth it?

That's a case-by-case kind of thing, and there are ways you could mitigate it...also costing a bit more code.

A symbol is also a bit more idiomatic than a string for this kind of thing.

File: config.rb (hunk 21 above)
[the remark this answers is not in the document]
Comment (furkansahin) [reply config.rb:67]: I have updated how I reference the api as suggested here

File: model/hetzner_host.rb [hunk 22]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+require_relative "../model"
+
+class HetznerHost < Sequel::Model
+  one_to_one :vm_host, key: :id
+
+  PROVIDER_NAME = "hetzner".freeze
```
Comment (fdr) [inline model/hetzner_host.rb:8]: String literals are automatically frozen given the `# frozen_string_literal: true` written into the top of the file
```suggestion
  PROVIDER_NAME = "hetzner"
```

**Review by fdr — COMMENTED** [review]
I'm basically good wth this, but I'm perplexed by the src/dst fields on Address. For one, what happens if we move failover IP(s) from one host to other hosts, and then delete/unsubscribe the first host?

I'm also puzzled because grepping for "src_vm_host" and "dst_vm_host" shows the association definitions, and the column...but are they traversed anywhere?

**Comment by furkansahin** [comment]
So, Hetzner has a pretty loose failover ip implementation.

1. You buy failover IP for a server
2. Even if you perform a failover, it's not listed under the new server, it sticks to original one.
3. If you deprovision the original server you have bought the failover ip for, even if it's assigned to a different server, the failover ip block also gets dropped.
Right now we don't have ip failover logic but we will, and we won't want to drop servers that has an ip ranged assigned elsewhere. That's why keeping src/dst colums is important.

File: prog/vm/host_nexus.rb (hunk 19 above)
Comment (fdr) [reply prog/vm/host_nexus.rb]: Seems fine, I mean, I guess interactively I write "hetzner" but in source I'll write the symbol. Most of the time. Easier to grep I guess?

**Comment by fdr** [comment]
Man that's brutal. I opened a support ticket to see what they have to say, though it doesn't look good IMO. Maybe there's an API call where the billing record (which I think is what this bizarre interpretation might be partially about)  can be moved as well.

**Comment by fdr** [comment]
Can you improve this patch's legibility by separating the implied IPv4 patch from the also-implied Hetzner API patch?

Also, given we (via support emails) understand the Hetzner situation a bit better with regard to the odd attachment IPs have to servers, let's rename "src" and "dst" because this has a lot of lexical clashing with tons of other network nomenclature. "billed_to" and "routed_to" would be better.

Which raises the additional issue of, embedding these extremely weird fields for Hetzner into the Address model is, well, strange. But maybe it can be left there anyway, if the cure (joining an extra table) is worse than the disease.

Another option is to not have "billed_to." Maybe I missed it, but I don't see how it's used in the program other than to record things, i.e. duplicative with Hetzner's database? That can be fine in some circumstances, but I don't see how that's true here: apparently, they do have the power to modify what billing line item the failover IP is charged to within certain network technology related constraints, so doubling-up bookkeeping is asking for the creation of discrepancy, and incurring that problem should come with a gratifying advantage.

**Comment by furkansahin** [comment]
I have removed the Hetzner api related parts and src_host_id. The dst_host_id is changed to routed_to_host_id. I will post a new pr for hetzner API.

File: prog/vm/host_nexus.rb [hunk 23]
```diff
@@ -1,18 +1,28 @@
 # frozen_string_literal: true
 
 class Prog::Vm::HostNexus < Prog::Base
-  subject_is :sshable, :vm_host
+  subject_is :vm_host
```
Comment (fdr) [inline prog/vm/host_nexus.rb:4]: That's odd...is the subject no longer sshable? Did this change sneak in from conflict resolution?

File: rhizome/lib/vm_setup.rb [hunk 24]
```diff
@@ -82,16 +77,20 @@ def hugepages(mem_gib)
 
   def interfaces
     r "ip netns add #{q_vm}"
-
```
Comment (fdr) [inline rhizome/lib/vm_setup.rb:85]: Seems a spurious deletion clogging up the patch hunks.

File: rhizome/lib/vm_setup.rb [hunk 25]
```diff
@@ -151,7 +150,29 @@ def parse_routes(routes)
...
+    local_ip = NetAddr::IPv4Net.parse(ip4_local)
+    vm, vetho, vethi = [vm_sub.to_s,
+      local_ip.network.to_s,
+      local_ip.next_sib.network.to_s]
+
+    r "ip addr add #{vetho}/32 dev vetho#{q_vm}"
+    r "ip route add #{vm} dev vetho#{q_vm}"
+    r "echo 1 > /proc/sys/net/ipv4/conf/vetho#{q_vm}/proxy_arp"
+
+    r "ip -n #{q_vm} link set dev lo up"
```
Comment (fdr) [inline rhizome/lib/vm_setup.rb:165]: Do we use the loopback interface? This affects more than just ipv4 routes, if we need it up, it sort of doesn't belong here.

**Review by fdr — COMMENTED** [review]
I have to think about a few more things, but this does help a lot in narrowing the inner-loop of evaluation/re-evaluation than the combined Hetzner patch (which doubles the amount of stuff going on, which is more 4x+ the scan/rescan work for the reviewer)

**Comment by fdr** [comment]
Can you write a commit message detailed enough that demonstrates how to try the patch out?

**Comment by fdr** [comment]
Also it would be nice if your editor would wrap the commit messages to be more legible when running `git log`:

```
commit 24487a4a7e87508ea377d491226cbef8a6860831
Author: Furkan Sahin <furkan@ubicloud.com>
Date:   Sun Jun 4 17:13:45 2023 +0200

    Support NDP Proxying
    
    We are running on a landscape that different hosting providers have different implementations on how they redirect traffic for IPv6 ranges.
    There are 2 main themes;
    1. For a given subnet, redirect the whole traffic according to prefix by skipping ndp. In this implementation, if we are assigned /64 subnet, all the packets in that range will be sent to the host.
    2. We are assigned /64 subnet, we split it to /79 chunks but when we need to send a packet to that /79 chunk, switch sends ns messages to resolve the MAC address. If it doesn't get any response, the package is dropped. That's why our 
host needs to respond to these messages via proxy ndp and we should teach it to respond a specific ip address even if it is under our range. This comes with a cost of not being able to use all the ip addresses in the range since we have t
o add them 1 by 1 but it is a hosting provider limitation and we cannot do much atm.
```

File: rhizome/lib/vm_setup.rb (hunk 25 above)
Comment (furkansahin) [reply rhizome/lib/vm_setup.rb]: This was there from the manual ip assignment days. While trying many different commands, it sneaked its way into here.

File: prog/vm/host_nexus.rb (hunk 23 above)
Comment (furkansahin) [reply prog/vm/host_nexus.rb]: yes, it sneaked in

**Comment by furkansahin** [comment]
@fdr I have made the changes you requested

**Comment by fdr** [comment]
> @fdr I have made the changes you requested

alright I will give it a try and probably you commit when you wake up

**Comment by fdr** [comment]
I tweaked the commit message. I removed the backslashes, since  Github (when viewing one commit) nor github really do anything with escaped markdown there; markdown is only good insomuch as it can increase legibility even in plaintext, and for conveniently setting the PR description which *is* rendered markdown.

**Review by fdr — APPROVED** [review]
I appreciate the instructions and the limitation about the tap device (which I vaguely knew but is nice to see repeated, maybe it would have been nice to see a mention of DHCP4's limitation in particular, but that's a debatable inclusion)
